book/python/using-connectors.md

```diff
@@ -0,0 +1,191 @@
+# Using Connectors As Sources & Sinks
+
+Wallaroo's connectors are an experimental feature that allows full customization of sources and sinks. These streams can be integrated into Wallaroo using libraries you're already familiar with in Python. A number of built-in connectors offer a quick way to get started and hook up common stream types and they're all available for easy customization.
+
+Care has been taken to keep the API simple and the runtime assumtions unobtrusive so it's easy to hook up the library to existing code. Read down below for more information on how this works.
+
+Connectors themselves are composed of to two parts. First is the entry in your application builder which explains how encoding and decoding of the content works as well as giving each connector instance a name and direction of flow (source or sink). Second is the script which is a local process which runs side-by-side with Wallaroo. We'll look at each part in turn in the sections below.
+
+## Using A Connector In Your Application
+
+In your application module's `application_setup` function, the application builder returns a description of your application in terms of pipelines. Connectors tap into this by allowing the declaration of source and sink connectors and wiring these up by name in your pipeline defitions.
+
+Here is an example of what this looks like:
+
+```python
+def application_setup(args):
+    ab = wallaroo.ApplicationBuilder("Celsius to Fahrenheit")
+
+    ab.source_connector("celsius_feed",
+        encoder=encode_feed,
+        decoder=decode_feed)
+
+    ab.sink_connector("fahrenheit_conversion",
+        encoder=encode_conversion,
+        decoder=decode_conversion)
+
+    ab.new_pipeline("Celsius Conversion", "celsius_feed")
+    ab.to(multiply)
+    ab.to(add)
+    ab.to_sink("fahrenheit_conversion")
+    return ab.build()
+```
+
+You can follow along in this example by going to [github](https://github.com/Wallaroo/wallaroo/tree/{{book.wallaroo_version}}/examples/python/celsius_connectors/).
+
+We have introduced two new application builder methods for declaring connectors: `source_connector` and `sink_connector`. These allow you to describe both ends of the connection so the connector script can encode or decode data in a way that's compatible with your application. Keeping these in one place helps ensure that it's easy to keep these in sync.
+
+In our example we can look at the celsius feed's encoder and decoder functions:
+
+```python
+@wallaroo.experimental.stream_message_encoder
+def encode_feed(data):
```

Comment:
So these functions always have to be defined regardless if we don't need to re-encode it? Does it make sense to eventually make these functions optional, allowing you to redefine them if you need to?

Reply:
Yes. I originally kept identity functions around but it's far better to be explicit. After writing a few apps, I found it hard to recall what kind of processing was done on inputs or outputs and having it spelled out makes it clear that it was intended rather than an omission.